### PR TITLE
Fixes #7: Support multi-line replies from Tor.

### LINKED ISSUE #7: Requesting multi-line replies caused premature `ProtocolError` exception
* **This issue is reproducible.**

# Overview

In some cases, Tor responds with data that spans multiple lines. It signifies this with a `+` as per section 2.2 of the Tor control spec:

> A command is either a single line containing a Keyword and arguments, or a multiline command whose initial keyword begins with +, and whose data section ends with a single "." on a line of its own.

Further, in section 2.3 of the spec:

> Multiple lines in a single reply from Tor to the controller are guaranteed to share the same status code.

One such command that you can issue to Tor to get a multi-line reply is the `GETINFO config-text` command.

`php-torcontrol` cannot handle such replies, prematurely throwing a `ProtocolError` on the second line of the response of a multi-line response.

# Steps to reproduce

1. Set up a Tor whose configuration file (`torrc`) contains more than single configuration line.
1. Start Tor.
1. Connect `php-torcontrol` to that Tor server.
1. Issue the `GETINFO config-text` command.
1. Observe premature error.

Code example to perform the above:

```php
function get_tor_config_text () {
    $tc = new \TorControl\TorControl();
    $tc->connect();
    $tc->authenticate();
    $response = $tc->executeCommand('GETINFO config-text');
}
```

# Expected results

The `$response` array contains all lines of the Tor reply, without triggering an error.

# Actual results

A `ProtocolError` exception thrown by `TorControl::executeCommand()` occurs.

# Other information

I have PHPUnit test cases along with a suggested patch to fix the issue, will send the pull request in a moment.

### PR DESCRIPTION
**Pull request by meitar**
This commit fixes an issue where a `ProtocolError` exception was
prematurely thrown before the end of a multi-line reply from Tor. The
issue was caused by ignoring the multi-line nature of certain replies
from Tor and always assuming single-line replies. This patch addresses
the issue by retaining state information about whether the current line
is a multi-line or single line reply.

This patch also fixes a typo in an error message (`Unknow`-> `Unknown`).

This patch includes unit tests that demonstrate the problem and the fix.

### CODE REVIEW

File: src/TorControl/TorControl.php [hunk 1]
```diff
@@ -247,37 +247,57 @@ public function executeCommand($cmd)
...
+                    }
                 }
 
-                $data[] = array(
-                    'code' => $code,
-                    'separator' => $separator,
-                    'message' => $message,
-                );
+
+                if ($multiline) {
```
Comment (dunglas) [inline src/TorControl/TorControl.php]: Could be simplified using a `else`.

File: src/TorControl/TorControl.php [hunk 2]
```diff
@@ -247,37 +247,57 @@ public function executeCommand($cmd)
...
+                    $data[] = array(
+                        'code' => $last_code,
+                        'separator' => $last_separator,
+                        'message' => $line
+                    );
+                } else {
+                    $data[] = array(
```
Comment (dunglas) [inline src/TorControl/TorControl.php]: Could be moved at the end of the `if (!$multiline)` block (line 284).

File: src/TorControl/TorControl.php [hunk 3]
```diff
@@ -247,37 +247,57 @@ public function executeCommand($cmd)
...
+                        'code' => $code,
+                        'separator' => $separator,
+                        'message' => $message
+                    );
+                }
 
                 if ($separator === ' ') {
```
Comment (dunglas) [inline src/TorControl/TorControl.php]: Please use yoda style (we follow Symfony CS).

File: tests/TorControl/Tests/TorControlTest.php [hunk 4]
```diff
@@ -82,4 +82,39 @@ public function testAuthenticate()
 
         $this->assertFalse($this->torControl->getOption('authmethod') === TorControl::AUTH_METHOD_NOT_SET);
     }
+
+    /**
+     * Test multiline replies.
+     */
+    public function testMultilineReplies()
+    {
+        $_cmds = array(
```
Comment (dunglas) [inline tests/TorControl/Tests/TorControlTest.php]: A var starting with `_` is not valid according to PSR CS.

File: tests/TorControl/Tests/TorControlTest.php [hunk 5]
```diff
@@ -82,4 +82,39 @@ public function testAuthenticate()
...
+        foreach ($_cmds as $cmd) {
+            $responses[] = $this->torControl->executeCommand($cmd);
+        }
+
+        // Assert we got the same number of responses entries as commands sent.
+        // If we do, and there are no exceptions, we can handle multi-line replies.
+        $this->assertEquals(count($_cmds), count($responses));
+        // Test to ensure multiline reply codes are correctly populated.
+        foreach ($responses[2] as $resp) {
+            $this->assertSame($resp['code'], '250');
```
Comment (dunglas) [inline tests/TorControl/Tests/TorControlTest.php]: The first parameter is the expected value and the second the value to test (should be inverted, apply for all other assertions).

**Comment by dunglas** [comment]
:+1: 
I just left some minor comments.

File: tests/TorControl/Tests/TorControlTest.php (hunk 4 above)
Comment (meitar) [inline tests/TorControl/Tests/TorControlTest.php]: Sorry, bad habits die hard.

**Comment by meitar** [comment]
Okay, I inverted the `if … else` conditions to simplify the execution path, makes it easier to read, I think. Also, I admit I'm not super familiar with the PSR or Symphony coding style guides but I did try to follow them, so let me know if I missed something.

**Comment by dunglas** [comment]
Thank you very much for your contribution @meitar

**Pull request merged**